TASK: Use new `translate()` method if possible

mhsdesign opened the pull request:
Followup to https://github.com/neos/neos-ui/pull/3804 which deprecates the use of the <i18n> component and the i18nRegistry, as this is now a global concept via the `translate` method.

**What I did**

Using various careful regex replacements and manual adjustments, commit by commit the `translate()` method was introduced.

As written here https://github.com/neos/neos-ui/pull/3804#issuecomment-2613992743 the new method does not handle user input gracefully, which is the reason why non static strings are still using the old translation methods to ensure nothing breaks here.

**How I did it**

**How to verify it**

Via the regex

```
translate\('(?!(Neos\.Neos:(Main|Inspector|Modules)|Neos\.Neos\.Ui:(Error|Main|PublishingDialog|SyncWorkspaceDialog)):[a-zA-Z._-]+')
```

All source files can be checked that each call to `translate()` now specifies the translation as fully qualified shorthand string.

Replacing the <i18n> component with a direct call to `translate()` might not always look correct as this leads to a missing `<span />` tag. Those cases have to be identified by eye.

<!--
If possible, a screenshot or a gif comparing the new and old behavior would be great.
-->

mhsdesign committed:
TASK: Replace translation fallback closure pattern with parameter substitution in new `translate` function

mhsdesign committed:
TASK: Replace simple uses of <I18n /> in new publishing code

mhsdesign committed:
TASK: Replace complex uses of <I18n /> in new publishing code

mhsdesign committed:
TASK: Dont pass I18nRegistry around in new publishing code

mhsdesign committed:
TASK: Replace simple usages of <I18n /> across the codebase

Using the regex

```
<I18n[ \n]+id="([^"]*)"[ \n]+fallback="([^"]*)"[ \n]*/>
```

and replacing it via

```
{translate('$1', '$2')}
```

To target

```
<I18n id="Neos.Neos:Main:cancel" fallback="Cancel"/>
```

And migrate it to

```
{translate('Neos.Neos:Main:cancel', 'Cancel')}
```

Manual adjustments, to adjust the imports and fix cases where the replaced code would have too many braces like here were fixed according to the linter

{someExpression && {translate(...)}}

mhsdesign committed:
TASK: Replace simple usages of `i18nRegistry` across the codebase

Using the regex

```
(?<=[ {])(this.props.)?i18nRegistry\.translate\('Neos.
```

and replacing it via

```
translate('Neos.
```

To target all shorthand translations like `Neos.Neos:...` or from Neos.Neos.Ui.

Also adjusted the imports and remove the injections of i18nRegistry if now obsolete

mhsdesign committed:
TASK: Replace usages of `i18nRegistry` without qualified package across the codebase

Using the regex

```
(?<=[ {])i18nRegistry\.translate\('([^']+)'\)
```

and replacing it via

```
translate('Neos.Neos:Main:$1')
```

To target all translations like `i18nRegistry.translate('copy')` and replace them via `translate('Neos.Neos:Main:copy')`

Also adjusted the imports and remove the injections of i18nRegistry if now obsolete

mhsdesign committed:
TASK: Replace usages of `i18nRegistry` with explicitly passed parameters across the codebase

to shorthand strings

Using the regex

```
(?<=[ {])(?:this.props.)?i18nRegistry\.translate\('([^']+)'\, '([^']+)'\, \{\}, '([^']+)'\, '([^']+)'\)
```

and replacing it via

```
translate('$3:$4:$1', '$2')
```

To target all translations like `i18nRegistry.translate('contentElementsSelected', 'content elements selected', {}, 'Neos.Neos.Ui', 'Main')` and replace them via `translate('Neos.Neos.Ui:Main:contentElementsSelected', 'content elements selected')`

Also adjusted the imports and remove the injections of i18nRegistry if now obsolete

mhsdesign committed:
TASK: Replace leftover usages of `i18nRegistry` to shorthand string if possible

oddly (multiline) formatted things could not be targeted via regex replace

mhsdesign committed:
TASK: Replace odd usage of `i18nRegistry` to shorthand string

The todo references this pr which fixed the return value of `i18nRegistry.translate` long ago:

https://github.com/neos/neos-ui/pull/2594

mhsdesign committed:
TASK: Provide default fallback value for translate() in selectBoxHelper

mhsdesign committed:
TASK: Jest tests by faking new translate() method correctly

mhsdesign committed:
TASK: Replace usages of <I18n /> with only one parameter

Also appends Neos.Neos:Main if necessary, which was forgotten in:

05fe5f8e214ec42da43f3d6c46ade2c027220e95

mhsdesign committed:
TASK: Replace leftover usages of <I18n /> if possible

markusguenther commented:
Wow, that’s a lot to take in! I’ll need some time to look into it. Thanks a bunch!

mhsdesign committed:
TASK: Fix jest test and fix `PublishDropDown` neos prop

The `routes` propType has been removed as the routes are accessed via neos.routes

mhsdesign committed:
PATCH: Små post 9.1 rebase adjustments

mhsdesign committed:
PATCH: Adjust to use `translate()` in DimensionSelectorOption and StringLength validator

mhsdesign committed:
FEATURE: Use typescript to validate that the `translate()` function is always invoked with constant strings by the pattern Package:Source:UnitId

mhsdesign commented:
Currently the typescript interface of `translate()` demands a second parameter, the fallback. In some instances we do provide a fallback but we only provided a fallback so far half the time approximately ~130 times. The other half we have no fallback.
Now to satisfy typescript i added an empty string for the cases where we dont have anything filled out yet. That will be ~130 needless empty strings added then :) The question is do we want to have fallbacks explicitly for new code - in that case typescript will force us to think about it (or you insert `''` to make it silent, or do we dont think that Neos is navigateable without translation framework and files.

A third solution could be to use translation based on the actual label already what flow supports but what we have not used too much. See `Translator::translateByOriginalLabel()`

Sebobo commented:
IMHO, the empty string is worse. If the translation fails, you don't see anything and might not even the notice the issue. F.e. when the label is next to an icon in a button. Seems harder to debug this way?

mhsdesign committed:
TASK: Remove empty string `""` as dummy fallback value from translation calls

As discussed https://github.com/neos/neos-ui/pull/3912#issuecomment-3654337034 the empty string rather disguises bugs as nothing is output to the frontend.

To ease debugging the translation address should be returned again automatically.

See also https://github.com/neos/neos-ui/pull/4083

mhsdesign commented:
Typescript fails here because we require the change https://github.com/neos/neos-ui/pull/4083 upmerged

mhsdesign committed:
PATCH: Ignore expected ts error in unit test
